File: leetcode/201-300/230-kth-smallest-element-in-a-bst/bench/kth_smallest_bst.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    long val;
    long left;
    long right;
} Node;

static Node *nodes;
static long node_count;

static long insert(long root, long v) {
    if (root == -1) {
        long idx = node_count;
        nodes[idx].val = v;
        nodes[idx].left = -1;
        nodes[idx].right = -1;
        node_count += 1;
        return idx;
    }
    if (v < nodes[root].val) {
        long l = insert(nodes[root].left, v);
        nodes[root].left = l;
    } else {
        long r = insert(nodes[root].right, v);
        nodes[root].right = r;
    }
    return root;
}

static long kth_smallest(long root, long k, long cap) {
    long *stack = malloc(cap * sizeof(long));
    long top = 0;
    long cur = root;
    long count = 0;
    long result = -1;
    while (cur != -1 || top > 0) {
        while (cur != -1) {
            stack[top++] = cur;
            cur = nodes[cur].left;
        }
        long node = stack[--top];
        count += 1;
        if (count == k) {
            result = nodes[node].val;
            break;
        }
        cur = nodes[node].right;
    }
    free(stack);
    return result;
}

int main(void) {
    long n = 3000;
    long queries = 140000;

    nodes = malloc(n * sizeof(Node));
    node_count = 0;

    long root = -1;
    long state = 12345;
    for (long i = 0; i < n; i++) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        root = insert(root, state);
    }

    long sink = 0;
    for (long q = 0; q < queries; q++) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        long k = 1 + (state % n);
        sink += kth_smallest(root, k, n);
    }
    printf("%ld\n", sink);
    free(nodes);
    return 0;
}

File: leetcode/201-300/230-kth-smallest-element-in-a-bst/bench/kth_smallest_bst.rs
struct Node {
    val: i64,
    left: i64,
    right: i64,
}

fn insert(nodes: &mut Vec<Node>, root: i64, v: i64) -> i64 {
    if root == -1 {
        let idx = nodes.len() as i64;
        nodes.push(Node {
            val: v,
            left: -1,
            right: -1,
        });
        return idx;
    }
    if v < nodes[root as usize].val {
        let l = insert(nodes, nodes[root as usize].left, v);
        nodes[root as usize].left = l;
    } else {
        let r = insert(nodes, nodes[root as usize].right, v);
        nodes[root as usize].right = r;
    }
    root
}

fn kth_smallest(nodes: &[Node], root: i64, k: i64) -> i64 {
    let mut stack: Vec<i64> = Vec::new();
    let mut cur = root;
    let mut count = 0i64;
    while cur != -1 || !stack.is_empty() {
        while cur != -1 {
            stack.push(cur);
            cur = nodes[cur as usize].left;
        }
        let node = stack.pop().unwrap();
        count += 1;
        if count == k {
            return nodes[node as usize].val;
        }
        cur = nodes[node as usize].right;
    }
    -1
}

fn main() {
    let n: i64 = 3000;
    let queries: i64 = 140000;

    let mut nodes: Vec<Node> = Vec::new();
    let mut root: i64 = -1;
    let mut state: i64 = 12345;
    for _ in 0..n {
        state = (state * 1103515245 + 12345) & 2147483647;
        root = insert(&mut nodes, root, state);
    }

    let mut sink: i64 = 0;
    for _ in 0..queries {
        state = (state * 1103515245 + 12345) & 2147483647;
        let k = 1 + (state % n);
        sink += kth_smallest(&nodes, root, k);
    }
    println!("{}", sink);
}
